pkg/apb/executor.go

```diff
@@ -32,11 +33,117 @@ import (
 	"k8s.io/api/core/v1"
 )
 
-// ExecuteApb - Runs an APB Action with a provided set of inputs
-func ExecuteApb(action string,
-	spec *Spec,
-	context *Context,
-	p *Parameters) (ExecutionContext, error) {
+// ExecutorAccessors - Accessors for Executor state.
+type ExecutorAccessors interface {
+	PodName() string
+	LastStatus() StatusMessage
+	ExtractedCredentials() *ExtractedCredentials
+}
+
+// ExecutorAsync - Main interface used for running APBs asynchronously.
+type ExecutorAsync interface {
+	Provision(*ServiceInstance) <-chan StatusMessage
+	Deprovision(instance *ServiceInstance) <-chan StatusMessage
+	Bind(instance *ServiceInstance, parameters *Parameters) <-chan StatusMessage
+	Unbind(instance *ServiceInstance, parameters *Parameters) <-chan StatusMessage
+	Update(instance *ServiceInstance) <-chan StatusMessage
+}
+
+// Executor - Composite executor interface.
+type Executor interface {
+	ExecutorAccessors
+	ExecutorAsync
+}
+
+type executor struct {
+	extractedCredentials *ExtractedCredentials
+	podName              string
+	lastStatus           StatusMessage
+	statusChan           chan StatusMessage
+	mutex                sync.Mutex
+}
+
+// NewExecutor - Creates a new Executor for running an APB.
+func NewExecutor() Executor {
+	exec := &executor{
+		statusChan: make(chan StatusMessage),
+		lastStatus: StatusMessage{State: StateNotYetStarted},
+	}
+	return exec
+}
+
+// PodName - Returns the name of the pod running the APB
+func (e *executor) PodName() string {
+	return e.podName
+}
+
+// LastStatus - Returns the last known status of the APB
+func (e *executor) LastStatus() StatusMessage {
+	return e.lastStatus
+}
+
+// ExtractedCredentials - Credentials extracted from the APB while running,
+// if they were discovered.
+func (e *executor) ExtractedCredentials() *ExtractedCredentials {
+	return e.extractedCredentials
+}
+
+func (e *executor) start() {
+	log.Debug("executor::start")
+	status := e.lastStatus
+	status.State = StateInProgress
+	e.lastStatus = status
+	e.statusChan <- status
```

Comment:
I'm confused by this, we copy the lastStatus into status, set its state, then set it back on lastStatus. Why isn't this just:

```golang
func (e *executor) start() {
    log.Debug("executor::start")
    e.lastStatus.State = StateInProgress
    e.statusChan <- e.lastStatus
}
```

Reply:
I think I was on autopilot here and it felt like a redux action, which is of course totally irrelevant here. I'll update :smile: